### ISSUE TITLE
Suggestion: TempoQR on TimeQuestions dataset

### ISSUE THREAD

**Issue by apoorvumang** [cmavro/TempoQR#1]
Hi, very interesting work!

I am one of the authors of the CronKGQA paper. Other than CronQuestions, I have been looking at other temporal KGQA datasets, one of which is TimeQuestions (https://exaqt.mpi-inf.mpg.de/). Since it is based on WikiData and not on a 'temporal KG', I processed the data provided by (https://github.com/zhenjia2017/EXAQT) using some heuristics to create a dataset somewhat similar in format to CronQuestions. You can find this processed dataset here: https://storage.googleapis.com/cronkgqa/timequestions.zip

I ran CronKGQA and TempoQR on TimeQuestions to the best of my abilities and got the following results:

![image (3)](https://user-images.githubusercontent.com/1957903/153710120-b8946e75-8d48-4563-b013-7f43f0eeffb8.png)

It would be great if you could also try out TempoQR on this dataset and see if you get different numbers. In my work, I am planning to report the numbers for TempoQR I obtained through my reproduction on TimeQuestions. If you think I am going wrong somewhere or you are able to get better results, I would be very happy to include those.

Thanks
Apoorv

**Comment by cmavro** [comment]
Hi Apoorv,

Thank you a lot for your interest and for providing us the additional dataset. I am currently doing the experimentation, but I have some comments/questions regarding the QA dataset.
1) The training size is somewhat smaller (6970 qs) than the one used in EXAQT (9708 qs), which might impact the overall performance of TempoQR and CronKGQA. 
2) Moreover, some questions do not contain answers (~300-400 have empty answer sets for train and test, respectively). Do you filter them out during the QA process?

Based on preliminary results with the above configuration, I am able to verify the reported numbers for TempoQR-soft. However, I am observing that it still works better than CronKGQA (using our code). I am using distill-bert as the LM and TComplEx embeddings, which achieve KG completion performance about 87% and 20% at Hits@1 during training and testing, respectively.  Is this similar to your experimentation? 
Furthermore, it looks like when we unfreeze the TKG embeddings, both TempoQR and CronKGQA are able to achieve much better performance (overlapping train/test entities or possibly low quality embeddings?).

I am providing these results [here](https://drive.google.com/drive/folders/10OuwA_saja-oVd-h27q7OTNfugEdjElS?usp=sharing) (let me know if you have access), while I continue the reproduction. It would be great if you have any feedback (e.g., hyper-parameter tuning), so that I can adapt the experimental setup accordingly. Thanks a lot!

**Comment by apoorvumang** [comment]
Thanks so much for the detailed analysis + TKG embeddings. [Here](https://github.com/cmavro/TempoQR/files/8058630/create_tq_tkg_and_questions_with_template.zip) is the script I used for creating the TKG-format dataset from EXAQT's data - there could be some improvement possible here (place notebook inside EXAQT repo's root dir before running)

1. I somehow missed this discrepancy. Yes according to the paper there should be 9k train questions not 6.7k. I have raised this issue in the EXAQT github repo (https://github.com/zhenjia2017/EXAQT/issues/2), hopefully the author responds soon
2. For questions without answer, I filtered them out during training. However I did not remove them during testing for fair comparison.

The full TKG is just union of the facts of all questions. Since there was no 'gold' entity linking there are a lot of noisy/incomplete annotations. This also leads some questions unanswerable since answer entity was not found in the neighbourhood of qn entities.

Unfreezing significantly boosts performance, maybe because of test/train entity set overlap? I haven't checked this but this is also my hypothesis.

Regarding hyperparameter tuning, I haven't tried that much myself. I think we can improve on the dataset processing side as well to improve results. Also lets hope the EXAQT authors are able to provide subgraphs for the remaining ~3k train questions.

Please do let me know if you find any bugs/anything interesting in the data, and what numbers you are able to get from TempoQR/CronKGQA. Thanks again.

**Comment by cmavro** [comment]
Thanks for the script and the additional feedback.
 1. I see, so in order to answer the 3k extra questions, we need to compute TKG embeddings from scratch for the expanded TKG. 
 2. Regarding the test evaluation: as I understand this, all methods (TempoQR/CronKGQA/EXAQT) cannot answer these questions. However, since TempoQR/CronKGQA operate on the whole KG, they may have better chances to tackle these questions if the answers were annotated in the whole KG (although not present in the subgraph). Am I correct? 
 
 I will provide the results for TempoQR-hard as well as the results when including all test questions. Let me know if you have the expanded TKG so I can repeat the experiments.

**Comment by apoorvumang** [comment]
Yes, methods operating on whole graphs should have a better chance. In fact, I checked the test set answer recall *over the full union KG* and found it to be around 87%, which is higher than the 72.4% recall reported by EXAQT since they test the recall on question specific subgraphs.

I will let you know if I am able to obtain annotations for the remaining train questions

**Comment by apoorvumang** [comment]
btw the processed dataset I provided + dataset creation script contains a bug which causes lots of questions to not have answers.

Questions with annotation 'Temp.Ans' don't always have time as answer. I do a filtering step where I remove all answers that are not time for questions with this annotation. This should not be done. This will result in around 180 more questions in test set having non-zero answer set.

**Comment by cmavro** [comment]
Could you provide us the updated QA dataset for a fair comparison?

**Comment by apoorvumang** [comment]
This is the version I'm using right now: https://storage.googleapis.com/cronkgqa/timequestions_v2.zip

I would suggest that you don't remove questions without answers for evaluation splits. In the dataset `__getitem__` you can do something like this 
```
if len(answers_arr) == 0:
    answers_single = 0
else:
    answers_single = random.choice(answers_arr)
```

**Comment by cmavro** [comment]
These are results I'm getting with the above configuration: [Resultsv2](https://docs.google.com/document/d/194zIPtJdUxreMw7qXESR83aBXWI4kXPqOeWOPQeB_dg/edit?usp=sharing).

**Comment by apoorvumang** [comment]
Thank you. btw the results I showed for CronKGQA were using a modified model. I will use the numbers provided by you for TempoQR

**Comment by zhenjia2017** [comment]
> 
If the answers are events, we also categorized these questions as Temp.Ans.

**Comment by zhenjia2017** [comment]
Thanks a lot for your interest in TimeQuestions!

**Comment by zhenjia2017** [comment]
> Hi Apoorv,
> 
> Thank you a lot for your interest and for providing us the additional dataset. I am currently doing the experimentation, but I have some comments/questions regarding the QA dataset.
> 
> 1. The training size is somewhat smaller (6970 qs) than the one used in EXAQT (9708 qs), which might impact the overall performance of TempoQR and CronKGQA.
> 2. Moreover, some questions do not contain answers (~300-400 have empty answer sets for train and test, respectively). Do you filter them out during the QA process?
> 
> Based on preliminary results with the above configuration, I am able to verify the reported numbers for TempoQR-soft. However, I am observing that it still works better than CronKGQA (using our code). I am using distill-bert as the LM and TComplEx embeddings, which achieve KG completion performance about 87% and 20% at Hits@1 during training and testing, respectively. Is this similar to your experimentation? Furthermore, it looks like when we unfreeze the TKG embeddings, both TempoQR and CronKGQA are able to achieve much better performance (overlapping train/test entities or possibly low quality embeddings?).
> 
> I am providing these results [here](https://drive.google.com/drive/folders/10OuwA_saja-oVd-h27q7OTNfugEdjElS?usp=sharing) (let me know if you have access), while I continue the reproduction. It would be great if you have any feedback (e.g., hyper-parameter tuning), so that I can adapt the experimental setup accordingly. Thanks a lot!

When doing the EXAQT process, we removed the questions without ground truth in their subgraphs. So the number of training subgraphs is smaller than questions in the data set. The reason that some questions have no answers is that in WikiData there is no answer for them. We do not filter them out during the process.

**Comment by cmavro** [comment]
Hi, thanks for the clarification.
